File: src/lib/qoi.ts
import { Bitmap } from "./bitmap";

type QoiHeader = {
  width: number;
  height: number;
  channels: number;
  colorspace: number;
  headerSize: number;
};

type QoiPixel = {
  r: number;
  g: number;
  b: number;
  a: number;
};

function uintArrayContains(
  str: Uint8ClampedArray,
  array: Uint8ClampedArray,
  offset: number = 0
) {
  if (str.length + offset > array.length) return false;

  for (let k = 0; k < str.length; k++) {
    if (str[k] !== array[k + offset]) return false;
  }

  return true;
}

function uintArrayReadInt32BE<T extends Uint8ClampedArray | Uint8Array>(
  array: T,
  offset: number
) {
  return (
    array[offset + 3] |
    (array[offset + 2] << 8) |
    (array[offset + 1] << 16) |
    (array[offset] << 24)
  );
}

export class QoiParser {
  private parseHeader(contents: Uint8ClampedArray): QoiHeader | null {
    const enc = new TextEncoder();
    const magic = new Uint8ClampedArray(enc.encode("qoif"));
    if (!uintArrayContains(magic, contents)) return null;

    // image width in pixels (BE)
    const width = uintArrayReadInt32BE(contents, 4);
    // image height in pixels (BE)
    const height = uintArrayReadInt32BE(contents, 8);
    // 3 = RGB, 4 = RGBA
    const channels = contents[12];
    // 0 = sRGB with linear alpha
    // 1 = all channels linear
    const colorspace = contents[12];

    //console.log(width, height, channels, colorspace);
    return { width, height, channels, colorspace, headerSize: 14 };
  }

  private parsePixels(header: QoiHeader, contents: Uint8ClampedArray) {
    const pixels = new Uint8ClampedArray(
      new ArrayBuffer(header.width * header.height * 4)
    );
    let prevArray: QoiPixel[] = Array.from({ length: 64 }, () => ({
      r: 0,
      g: 0,
      b: 0,
      a: 0,
    }));

    const indexPosition = (pixel: QoiPixel) =>
      (pixel.r * 3 + pixel.g * 5 + pixel.b * 7 + pixel.a * 11) % 64;

    let p = header.headerSize;
    let idx = 0;

    const QOI_OP_RGB = 0b11111110;
    const QOI_OP_RGBA = 0b11111111;

    const setPixel = (pixel: QoiPixel) => {
      pixels[idx] = pixel.r;
      pixels[idx + 1] = pixel.g;
      pixels[idx + 2] = pixel.b;
      pixels[idx + 3] = pixel.a;

      idx += 4;
    };

    const wrapUint8 = (n: number) => {
      let x = n;
      while (x < 0) x += 256;
      return x % 256;
    };

    const opRgba = (rgbOnly: boolean, prevPixel: QoiPixel) => {
      const r = contents[p];
      const g = contents[p + 1];
      const b = contents[p + 2];
      const a = rgbOnly ? prevPixel.a : contents[p + 3];

      setPixel({ r, g, b, a });
      p = rgbOnly ? p + 3 : p + 4;
    };

    const opIndex = (tag: number) => {
      const index = tag & 0x3f;
      const pixel = prevArray[index];
      setPixel(pixel);
    };

    const opDiff = (tag: number, prevPixel: QoiPixel) => {
      const dr = ((tag >> 4) & 0x3) - 2;
      const dg = ((tag >> 2) & 0x3) - 2;
      const db = (tag & 0x3) - 2;

      const pixel: QoiPixel = {
        r: wrapUint8(prevPixel.r + dr),
        g: wrapUint8(prevPixel.g + dg),
        b: wrapUint8(prevPixel.b + db),
        a: prevPixel.a,
      };

      setPixel(pixel);
    };

    const opLuma = (tag: number, prevPixel: QoiPixel) => {
      const payload = contents[p++];

      const diffGreen = (tag & 0x3f) - 32;
      const drDg = ((payload >> 4) & 0x0f) - 8;
      const dbDg = (payload & 0x0f) - 8;

      const pixel: QoiPixel = {
        r: wrapUint8(diffGreen + drDg + prevPixel.r),
        g: wrapUint8(diffGreen + prevPixel.g),
        b: wrapUint8(diffGreen + dbDg + prevPixel.b),
        a: prevPixel.a,
      };

      setPixel(pixel);
    };

    const opRun = (tag: number, prevPixel: QoiPixel) => {
      const count = (tag & 0x3f) + 1;
      for (let k = 0; k < count; k++) setPixel(prevPixel);
    };

    const endSequence = Uint8Array.from([0, 0, 0, 0, 0, 0, 0, 1]);
    const detectEndSequence = (p: number) => {
      for (let k = 0; k < endSequence.length; k++) {
        if (contents[p + k] !== endSequence[k]) return false;
      }

      return true;
    };

    while (p < contents.length) {
      // Retrieving last pixel
      const prevPixel: QoiPixel =
        idx === 0
          ? { r: 0, g: 0, b: 0, a: 255 }
          : {
              r: pixels[idx - 4],
              g: pixels[idx - 3],
              b: pixels[idx - 2],
              a: pixels[idx - 1],
            };
      prevArray[indexPosition(prevPixel)] = prevPixel;

      // Parsing the operation tag
      const tag = contents[p++];
      if (tag === 0 && detectEndSequence(p - 1)) {
        break;
      }

      if (tag === QOI_OP_RGB || tag === QOI_OP_RGBA) {
        opRgba(tag === QOI_OP_RGB, prevPixel);
        continue;
      }

      const tagOp = tag >> 6;
      switch (tagOp) {
        // QOI_OP_INDEX
        case 0b00:
          opIndex(tag);
          break;

        // QOI_OP_DIFF
        case 0b01:
          opDiff(tag, prevPixel);
          break;

        // QOI_OP_LUMA
        case 0b10:
          opLuma(tag, prevPixel);
          break;

        // QOI_OP_RUN
        case 0b11:
          opRun(tag, prevPixel);
          break;
      }
    }

    return pixels;
  }

  parse(buffer: Uint8ClampedArray): Bitmap | null {
    const header = this.parseHeader(buffer);
    if (!header) return null;

    const pixels = this.parsePixels(header, buffer);
    if (!pixels) return null;

    return new Bitmap(header.width, header.height, pixels);
  }
}

File: rs-wasm-encoders/src/encoder/qoi.rs
use super::image::{Image, ImageCodec, ImageFormat, Result};
use bytes::{BufMut, BytesMut};

#[derive(Copy, Clone, Debug, PartialEq)]
struct QoiPixel {
    r: u8,
    g: u8,
    b: u8,
    a: u8,
}

const QOI_OP_RGB: u8 = 0b11111110;
//const QOI_OP_RGBA: u8 = 0b11111111;
const QOI_OP_INDEX: u8 = 0b00000000;
const QOI_OP_DIFF: u8 = 0b01000000;
const QOI_OP_LUMA: u8 = 0b10000000;
const QOI_OP_RUN: u8 = 0b11000000;

enum QoiOp {
    Rgb,
    //Rgba,
    Index,
    Diff,
    Luma,
    Run,
}

pub struct Qoi<'a> {
    image: &'a Image,
}
impl<'a> Qoi<'a> {
    pub fn from_image(image: &'a Image) -> Qoi<'a> {
        Qoi { image }
    }

    fn qoi_header(&self, contents: &mut BytesMut) {
        //  char magic[4]; // magic bytes "qoif"
        contents.put(&b"qoif"[..]);
        //  uint32_t width; // image width in pixels (BE)
        contents.put_u32(self.image.width() as u32);
        //  uint32_t height; // image height in pixels (BE)
        contents.put_u32(self.image.height() as u32);
        //  uint8_t channels; // 3 = RGB, 4 = RGBA
        contents.put_u8(3);
        //  uint8_t colorspace; // 0 = sRGB with linear alpha
        //  // 1 = all channels linear
        contents.put_u8(1);
    }

    fn encode_op_index(
        contents: &mut BytesMut,
        pixel: &QoiPixel,
        prev_array: &[QoiPixel; 64],
        pixel_idx: usize,
    ) -> bool {
        if pixel == &prev_array[pixel_idx] {
            contents.put_u8(QOI_OP_INDEX | pixel_idx as u8);
            true
        } else {
            false
        }
    }

    fn encode_op_diff(contents: &mut BytesMut, pixel: &QoiPixel, prev_pixel: &QoiPixel) -> bool {
        use std::num::Wrapping;

        let cdiff = |cur: u8, prev: u8| {
            let d = Wrapping(cur as i8) - Wrapping(prev as i8) + Wrapping(2);

            match d.0 {
                0..=3 => Some(d.0 as u8),
                _ => None,
            }
        };

        let rd = cdiff(pixel.r, prev_pixel.r);
        let rg = cdiff(pixel.g, prev_pixel.g);
        let rb = cdiff(pixel.b, prev_pixel.b);

        if let (Some(rd), Some(gd), Some(bd)) = (rd, rg, rb) {
            contents.put_u8(QOI_OP_DIFF | (rd << 4) | (gd << 2) | bd);
            return true;
        }

        false
    }

    fn encode_op_luma(contents: &mut BytesMut, pixel: &QoiPixel, prev_pixel: &QoiPixel) -> bool {
        use std::num::Wrapping;

        let green_diff = Wrapping(pixel.g as i8) - Wrapping(prev_pixel.g as i8);
        if (-32..=31).contains(&green_diff.0) {
            let rdiff =
                Wrapping(pixel.r as i8) - Wrapping(prev_pixel.r as i8) - green_diff + Wrapping(8i8);
            let bdiff =
                Wrapping(pixel.b as i8) - Wrapping(prev_pixel.b as i8) - green_diff + Wrapping(8i8);
            let range = 0..16;

            if range.contains(&rdiff.0) && range.contains(&bdiff.0) {
                contents.put_u8(QOI_OP_LUMA | (green_diff.0 + 32) as u8);
                contents.put_u8((rdiff.0 as u8) << 4 | bdiff.0 as u8);
                return true;
            }
        }

        false
    }

    fn encode_op_run(
        contents: &mut BytesMut,
        pixel: &QoiPixel,
        prev_pixel: &QoiPixel,
        prev_pixel_run_cnt: &mut u8,
    ) -> bool {
        if (pixel != prev_pixel || *prev_pixel_run_cnt >= 62) && *prev_pixel_run_cnt > 0 {
            contents.put_u8(QOI_OP_RUN | (*prev_pixel_run_cnt - 1));
            *prev_pixel_run_cnt = 0;
        }

        if pixel == prev_pixel {
            *prev_pixel_run_cnt += 1;
            return true;
        }

        false
    }

    fn encode_op_run_final(contents: &mut BytesMut, prev_pixel_run_cnt: &mut u8) {
        if *prev_pixel_run_cnt > 0 {
            contents.put_u8(QOI_OP_RUN | *prev_pixel_run_cnt);
            *prev_pixel_run_cnt = 0;
        }
    }

    fn encode_op_rgb(contents: &mut BytesMut, pixel: &QoiPixel) -> bool {
        contents.put_u8(QOI_OP_RGB);
        contents.put_u8(pixel.r);
        contents.put_u8(pixel.g);
        contents.put_u8(pixel.b);

        true
    }

    fn qoi_encode(&self, contents: &mut BytesMut) {
        let mut prev_array = [QoiPixel {
            r: 0,
            g: 0,
            b: 0,
            a: 0,
        }; 64];

        let index_position = |pixel: &QoiPixel| {
            ((pixel.r as u32 * 3 + pixel.g as u32 * 5 + pixel.b as u32 * 7 + pixel.a as u32 * 11)
                % 64) as usize
        };

        let mut i = 0;
        let pixel_buffer = &self.image.pixels();
        let mut prev_pixel_ref: Option<QoiPixel> = None;
        let ops = [
            QoiOp::Run,
            QoiOp::Index,
            QoiOp::Diff,
            QoiOp::Luma,
            QoiOp::Rgb,
        ];
        let mut prev_pixel_run_cnt = 0u8;

        while i < pixel_buffer.len() {
            let r = pixel_buffer[i];
            let g = pixel_buffer[i + 1];
            let b = pixel_buffer[i + 2];

            let pixel = QoiPixel { r, g, b, a: 255 };
            let pixel_idx = index_position(&pixel);

            for op in ops.iter() {
                match *op {
                    QoiOp::Run => {
                        if let Some(prev_pixel) = prev_pixel_ref {
                            if Qoi::encode_op_run(
                                contents,
                                &pixel,
                                &prev_pixel,
                                &mut prev_pixel_run_cnt,
                            ) {
                                break;
                            }
                        }
                    }
                    QoiOp::Index => {
                        if Qoi::encode_op_index(contents, &pixel, &prev_array, pixel_idx) {
                            break;
                        }
                    }
                    QoiOp::Diff => {
                        if let Some(prev_pixel) = prev_pixel_ref {
                            if Qoi::encode_op_diff(contents, &pixel, &prev_pixel) {
                                break;
                            }
                        }
                    }
                    QoiOp::Luma => {
                        if let Some(prev_pixel) = prev_pixel_ref {
                            if Qoi::encode_op_luma(contents, &pixel, &prev_pixel) {
                                break;
                            }
                        }
                    }
                    QoiOp::Rgb => {
                        if Qoi::encode_op_rgb(contents, &pixel) {
                            break;
                        }
                    } // _ => {
                      //     panic!("QOI operation not implemented!");
                      // }
                }
            }

            prev_array[pixel_idx] = pixel;
            prev_pixel_ref = Some(pixel);
            i += 3;
        }

        Qoi::encode_op_run_final(contents, &mut prev_pixel_run_cnt);

        // QOI_OP_STREAM_END
        for _i in 0..7 {
            contents.put_u8(0);
        }
        contents.put_u8(1);
    }
}

impl<'a> ImageFormat<'a> for Qoi<'a> {
    fn format_name() -> &'static str {
        "Qoi"
    }

    fn get_raw_image(&self) -> &'a Image {
        self.image
    }
}

impl<'a> ImageCodec<'a> for Qoi<'a> {
    fn serialize(&self) -> Result<Vec<u8>> {
        let mut contents = BytesMut::with_capacity(1024);
        self.qoi_header(&mut contents);
        self.qoi_encode(&mut contents);

        let c: Vec<u8> = Vec::from(contents.as_ref());
        Ok(c)
    }
}
